Addition of Imperal Gallons for UK users

asdfghjkai opened the pull request:
In the UK, we measure fuel consumption in imperial miles per gallon, but use litres to fill. The existing conditional statement will only allow imperal gallons if the fuelUnit == gal. This may work in most parts of the world, however for us strange brits using litres this results in iffy readings. This is a corrective fix to resolve this.

Has been tested against manual conversions - any feedback appreciated.

asdfghjkai committed:
Addition of Imperal Gallons for UK users

In the UK, we measure fuel consumption in imperial miles per gallon, but use litres to fill. The existing conditional statement will only allow imperal gallons if the fuelUnit == gal. This may work in most parts of the world, however for us strange brits using litres this results in iffy readings. This is a corrective fix to resolve this.

JoshStark commented:
Hi, thanks for the PR. 

I’ve not looked at this in quite a while but doesn’t the application already take into account the difference between Imperial and US gallons? When I wrote this I did it with imperial gallons in mind (I’m English) but last year a bug was brought to light where mpg was not being calculated correctly when the user’s fuel unit was Gal (US). 

The consumption converter doesn’t distinguish “mpg” and “mpg(Imperial)” because for the sake of conversion it checks the user’s specific fuel unit before converting. “mpg” is an interchangeable term. 

To be clear, when the app references “gallons” it is implicitly imperial. For US gallons it explicitly states so. 

Can you illustrate the issue you were seeing please? I think regarding this change it would add in an explicit new consumption metric but functionally would be the same as it currently is.

asdfghjkai commented:
Hi JoshStark - cheers for getting back to me.

Yes, so the issue can be easily shown in this little code snippet [here](https://github.com/linuxserver/Clarkson/blob/6d1f9a28246a78469c40e41efdcea10b52d89792/src/app/services/conversion/consumption-converter.ts)

`
if ('mpg' === consumptionUnit) {
            const fuelUnitAsGallons = ('gal' === fuelUnit ? fuelUnitConversion.toGallons() : fuelUnitConversion.toGallonsUS());
            return Math.round((distanceUnitConversion.toMiles() / fuelUnitAsGallons) * 100) / 100;
        }
`

Setup as a UK user, my consumptionUnit is mpg, but my fuel unit is litres not gal. As such, the lambda expression evaluates to using fuelUnitConversion.toGallonsUS(), resulting in the incorrect result.

One possible (and better for integration long term) alternative solution I thought of after submiting the PR is to instead expand the condition of the lambda statement to be either 'gal' OR 'l'. This would achieve the same goal.

eg have soemthing like
`
const fuelUnitAsGallons = ('gal' === fuelUnit || 'l' === fuelUnit ? fuelUnitConversion.toGallons() : fuelUnitConversion.toGallonsUS());
`

Or even better
`
const fuelUnitAsGallons = ('gal (US)' === fuelUnit ? fuelUnitConversion.toGallonsUS() : fuelUnitConversion.toGallons());
`


This would be better. Opinions would be good though, and hopefully I've explained this error properly.

JoshStark commented:
Ah, now I see. I hadn’t taken into account litres. I think your second suggestion is the best option. If you update the PR to make the statement as such I don’t see why I can’t accept it.

asdfghjkai committed:
Enabling accurate mpg calculation for UK users

Allowing accurate conversions without any revision to database contents, allowing in place upgrade for current users

asdfghjkai commented:
There we go - change made. Will be glad once this is brought into the docker image :) Cheers for the open discussion.

JStark-Sports commented:
No worries, thanks for taking the time to make the change.

JoshStark commented:
No worries. That looks good.
